keps/sig-network/3726-standard-application-protocols/README.md

```diff
@@ -0,0 +1,236 @@
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+There are cases where implementations implement different things for the same application protocol names. That already causes(and can cause more in the future) issues for certain implementations to interoperate. (see example with GKE and Istio under [Motivation](#motivation))
+
+This KEP proposes to declare standard Kubernetes protocol names for common protocols that are not IANA service names.
+
+
+## Motivation
+
+The lack of direct support for specifying application protocols for ports and the widespread use of implementation-specific annotations to mitigate it had led k8s to add the `AppProtocol` field to the port spec.
+
+While this is a good solution - we never came with recommended standards other than [IANA standard service names](https://www.iana.org/assignments/service-names) or a domain prefixed protocol.
+
+This loose definition has led us to have instances where implementations do different things for common protocols.
+
+One good example is `HTTP2`. 
+* In GKE you can use `appProtocol: HTTP2` and it will describe HTTP2 over TLS (https://cloud.google.com/kubernetes-engine/docs/how-to/secure-gateway#load-balancer-tls).
+* While in Istio it will be h2c (HTTP2 over cleartext).
+
+That creates a problem where users with GKE and Istio in their cluster can have very different behaviors for the same `appProtocol` value.
+
+
+
+### Goals
+
+* Build consensus around how common (non IANA service names) should be implemented
+* Help the broader community and specifically implementations to interoperate better
+* Provide short and clear documentation for how appProtocol values should be implemented
+  * Update the appProtocol user documentation respectively
+
+
+### Non-Goals
+
+* Validate appProtocol values
+* Monitor `appProtocol implementations
+
+
+## Proposal
+
+Kubernetes supports the `appProtocol` field to provide a way to specify an application protocol for each Service port.
+
+The [documentation](https://kubernetes.io/docs/concepts/services-networking/service/#application-protocol) for this field says that:
+
+```Values should either be IANA standard service names or domain prefixed names such as mycompany.com/my-custom-protocol.```
+
+This KEP proposes to declare standard Kubernetes protocol names for common protocols that are not IANA standard service names.
+
+Those common protocols will be well defined strings prefixed with ‘k8s.io’. 
+`k8s.io/h2c` as an example.
+
+### list of protocols
```

Comment:
Especially h2 vs h2c, and why we even need h2 when TLS already provides a mechanism for negotiating protocol

Reply:
In this case the relevant RFC includes clear definitions of both h2 and h2c (https://www.rfc-editor.org/rfc/rfc7540#section-3.1), I think it makes sense to copy those concepts and link out to the RFC. Although there is a mechanism for negotiating protocol, at least some implementations also provide a way to explicitly specify h2 or h2c.

Reply:
I guess a broader topic is - what happens when there is limited support.

For example, Istio cannot implement something that says "This is HTTPS, but it must be HTTP/2". But we could do "HTTPS", which is probably good enough for 99% of the users declaring "H2"... but then for that 1% of users, we didn't meet their demand (used HTTP/1 when they wanted h2). Alternatively, we could decide "This is H2, we can't always give that behavior, so we ignore it".

Hope that makes sense...

Reply:
In my opinion this is a good reason to make this a list of protocols supported by the underlying application. That feels like a follow up though. The scope of this KEP is really just trying to fill gaps in IANA Standard Service names which feels like it should be relatively non-controversial, even though we admit that follow up efforts will be needed to solve the larger problem.

Reply:
I didn't follow the example or the argument for a list, can you do it again but slower?

Reply:
I think I caused a confusion. 

This line is **"List of supported/standard protocols"** and it is **not** about changing the field to a list (which is suggested later in the followup work section)

I will rename this line to supported protocols.

To answer your question @thockin ; The idea is to declare standard protocol names for common protocols that are not IANA service names. That way in the current version of the field, implementations (istio/GKE etc etc) that implement non IANA service names like `http2` will be able to coexist. (without these standards they would need `gke.io/http2` and `istio.io/http2` and since `AppProtocol` is a string field now it can have only one of the options)

Probably for that reason most of the implementations support non prefixed app protocol names that are not IANA (and that should not be the case) - [istio as an example](https://github.com/istio/istio/blob/master/pkg/config/protocol/instance.go#L65-L94)

There is a suggested followup work to turn this field to a list instead of a string and truly support implementations to interoperate. (lets say two different impl specific prefixed protocols OR one impl specific prefixed and one not prefixed etc)